File: src/kernel/settings_navigation.py
"""Settings information architecture shared with API clients."""

from src.kernel.schemas.setting import SettingCategory, SettingItem, SettingsNavigationGroup

SETTINGS_NAVIGATION = (
    SettingsNavigationGroup(
        id="experience", categories=[SettingCategory.FRONTEND, SettingCategory.SESSION]
    ),
    SettingsNavigationGroup(
        id="intelligence",
        categories=[SettingCategory.AGENT, SettingCategory.LLM, SettingCategory.SCHEDULED_TASK],
    ),
    SettingsNavigationGroup(
        id="memory",
        categories=[
            SettingCategory.MEMORY,
            SettingCategory.MEMORY_EMBEDDING,
            SettingCategory.MEMORY_SEARCH,
            SettingCategory.MEMORY_STORAGE,
        ],
    ),
    SettingsNavigationGroup(
        id="capabilities",
        categories=[
            SettingCategory.TOOLS,
            SettingCategory.SKILLS,
            SettingCategory.SANDBOX,
            SettingCategory.FILE_UPLOAD,
            SettingCategory.DOCUMENT_PARSE,
            SettingCategory.AUDIO_TRANSCRIPTION,
        ],
    ),
    SettingsNavigationGroup(
        id="access",
        categories=[
            SettingCategory.USER,
            SettingCategory.SECURITY,
            SettingCategory.OAUTH,
            SettingCategory.CAPTCHA,
            SettingCategory.EMAIL,
        ],
    ),
    SettingsNavigationGroup(
        id="infrastructure",
        categories=[
            SettingCategory.MONGODB,
            SettingCategory.REDIS,
            SettingCategory.CHECKPOINT,
            SettingCategory.LONG_TERM_STORAGE,
            SettingCategory.S3,
            SettingCategory.TRACING,
        ],
    ),
)


def build_settings_navigation(
    settings: dict[str, list[SettingItem]],
) -> list[SettingsNavigationGroup]:
    """Only expose categories present after the service's permission filtering."""
    result = []
    for group in SETTINGS_NAVIGATION:
        categories = [category for category in group.categories if settings.get(category.value)]
        if categories:
            result.append(SettingsNavigationGroup(id=group.id, categories=categories))
    return result

File: frontend/src/components/panels/settingsNavigation.ts
import type {
  SettingCategory,
  SettingItem,
  SettingsNavigationGroup,
} from "../../types";
import type { VisibleCategory } from "./settingsPanelGrouping";

export const SETTINGS_NAV_GROUPS: {
  id: string;
  categories: SettingCategory[];
}[] = [
  { id: "experience", categories: ["frontend", "session"] },
  { id: "intelligence", categories: ["agent", "llm", "scheduled_task"] },
  {
    id: "memory",
    categories: [
      "memory",
      "memory_embedding",
      "memory_search",
      "memory_storage",
    ],
  },
  {
    id: "capabilities",
    categories: [
      "tools",
      "skills",
      "sandbox",
      "file_upload",
      "document_parse",
      "audio_transcription",
    ],
  },
  {
    id: "access",
    categories: ["user", "security", "oauth", "captcha", "email"],
  },
  {
    id: "infrastructure",
    categories: [
      "mongodb",
      "redis",
      "checkpoint",
      "long_term_storage",
      "s3",
      "tracing",
    ],
  },
];

export function buildSettingsNavigation(
  visible: VisibleCategory[],
  navigation: SettingsNavigationGroup[] = SETTINGS_NAV_GROUPS,
) {
  return navigation
    .map(({ id, categories }) => ({
      id,
      categories: categories.flatMap((category) =>
        visible.filter((entry) => entry.category === category),
      ),
    }))
    .filter((group) => group.categories.length > 0);
}

export function filterSettings(
  items: SettingItem[],
  options: {
    query: string;
    includeAdmin?: boolean;
    category: SettingCategory;
    subcategory: string | null;
    isVisible: (setting: SettingItem) => boolean;
    translate: (key: string) => string;
    categoryLabels: Partial<Record<SettingCategory, string>>;
    subcategoryLabels: Record<string, string>;
  },
) {
  const query = options.query.trim().toLocaleLowerCase();
  return items.filter((setting) => {
    if (
      (!options.includeAdmin && setting.frontend_visible === false) ||
      !options.isVisible(setting)
    )
      return false;
    if (!query)
      return (
        setting.category === options.category &&
        (options.subcategory === null ||
          (setting.subcategory || "") === options.subcategory)
      );
    return [
      setting.key,
      setting.description,
      options.translate(setting.description),
      options.categoryLabels[setting.category],
      options.subcategoryLabels[setting.subcategory],
    ].some((value) => value?.toLocaleLowerCase().includes(query));
  });
}
